src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,691 @@
+
+    for (const auto& t : spendable.saplingNoteEntries) {
+        saplingOutPoints.push_back(t.op);
+        saplingNotes.push_back(t.note);
+
+        libzcash::SaplingExtendedSpendingKey saplingKey;
+        assert(wallet.GetSaplingExtendedSpendingKey(t.address, saplingKey));
+        saplingKeys.push_back(saplingKey);
+
+        sum += t.note.value();
+        if (sum >= targetAmount) {
+            break;
+        }
+    }
+
+    // Fetch Sapling anchor and witnesses, and Orchard Merkle paths.
+    uint256 anchor;
+    std::vector<std::optional<SaplingWitness>> witnesses;
+    std::vector<std::pair<libzcash::OrchardSpendingKey, orchard::SpendInfo>> orchardSpendInfo;
+    {
+        LOCK(wallet.cs_wallet);
+        if (!wallet.GetSaplingNoteWitnesses(saplingOutPoints, anchorConfirmations, witnesses, anchor)) {
+            // This error should not appear once we're nAnchorConfirmations blocks past
+            // Sapling activation.
+            return TransactionBuilderResult("Insufficient Sapling witnesses.");
+        }
+        if (builder.GetOrchardAnchor().has_value()) {
+            orchardSpendInfo = wallet.GetOrchardSpendInfo(spendable.orchardNoteMetadata, builder.GetOrchardAnchor().value());
+        }
+    }
+
+    // Add Orchard spends
+    for (size_t i = 0; i < orchardSpendInfo.size(); i++) {
+        auto spendInfo = std::move(orchardSpendInfo[i]);
+        if (!builder.AddOrchardSpend(
+            std::move(spendInfo.first),
+            std::move(spendInfo.second)))
+        {
+            return TransactionBuilderResult(
+                strprintf("Failed to add Orchard note to transaction (check %s for details)", GetDebugLogPath())
+            );
+        }
+    }
+
+    // Add Sapling spends
+    for (size_t i = 0; i < saplingNotes.size(); i++) {
+        if (!witnesses[i]) {
+            return TransactionBuilderResult(strprintf(
+                "Missing witness for Sapling note at outpoint %s",
+                spendable.saplingNoteEntries[i].op.ToString()
+            ));
+        }
+
+        builder.AddSaplingSpend(saplingKeys[i].expsk, saplingNotes[i], anchor, witnesses[i].value());
+    }
+
+    // Add outputs
+    for (const auto& r : payments.GetResolvedPayments()) {
+        std::visit(match {
+            [&](const CKeyID& keyId) {
+                builder.AddTransparentOutput(keyId, r.amount);
+            },
+            [&](const CScriptID& scriptId) {
+                builder.AddTransparentOutput(scriptId, r.amount);
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                builder.AddSaplingOutput(
+                        r.isInternal ? internalOVK : externalOVK, addr, r.amount,
+                        r.memo.has_value() ? r.memo.value().ToBytes() : Memo::NoMemo().ToBytes());
+            },
+            [&](const libzcash::OrchardRawAddress& addr) {
+                builder.AddOrchardOutput(
+                        externalOVK, addr, r.amount,
```

Comment:
Should this not also be checking `r.isInternal` (as change can go to an Orchard receiver)?